asv/commands/run.py

```diff
@@ -229,29 +230,36 @@ def run(cls, conf, range_spec=None, steps=None, bench=None, parallel=1,
                             failed = value is None or (isinstance(value, list) and None in value)
 
                             if skip_failed and failed:
-                                skipped_benchmarks.add(key)
+                                skipped_benchmarks[result.env_name].add(key)
                             if skip_successful and not failed:
-                                skipped_benchmarks.add(key)
+                                skipped_benchmarks[result.env_name].add(key)
                 except IOError:
                     pass
 
             for env in environments:
                 for bench in benchmarks:
-                    if bench in skipped_benchmarks:
+                    if bench in skipped_benchmarks[env.name]:
                         log.step()
 
-            if not set(six.iterkeys(benchmarks)).difference(skipped_benchmarks):
-                continue
-
             if commit_hash:
                 log.info(
                     "For {0} commit hash {1}:".format(
                         conf.project, commit_hash[:8]))
 
             with log.indent():
                 for subenv in util.iter_chunks(environments, parallel):
-                    log.info("Building for {0}".format(
-                        ', '.join([x.name for x in subenv])))
+
+                    subenv_name = ', '.join([x.name for x in subenv])
```

Comment:
I think the right thing to do here is to do
```
active_environments = [env for env in environments if set(six.iterkeys(benchmarks)).difference(skipped_benchmarks[env.name])]

if not active_environments:
    continue

if commit_hash:
...
for subenv in util.iter_chunks(active_environments, parallel):
...
```

Reply:
I agree, this is cleaner - done!